functions/src/image/save-valid-image/save-valid-image.ts

```diff
@@ -0,0 +1,66 @@
+import * as functions from 'firebase-functions'
+import { db } from '../../index'
+import { Collections } from '../../model/collections'
+import { Image, ImageStatus, Mask } from '../../model/image'
+import { SaveValidImageRequest } from './model/save-image-request'
+
+const SAVED_MASKS_TO_COMPLETE = 4
+
+export const saveValidImage = functions
+  .region('europe-west1')
+  .https.onCall(async (data: SaveValidImageRequest, context) => {
+    if (!context.auth) {
+      throw new functions.https.HttpsError('unauthenticated', 'User must be authenticated.')
+    }
+
+    const labellerId = context.auth.uid
+    const { imageId, maskName } = data
+
+    try {
+      await db.runTransaction(async (transaction) => {
+        const imageRef = db.collection(Collections.IMAGES).doc(imageId)
+        const imageDoc = await transaction.get(imageRef)
+        const image = imageDoc.data() as Image
+
+        validateCanSaveValidImage(image, labellerId)
+
+        const labellers: string[] = [...(image.labellers ?? []), labellerId]
+        const mask: Mask = {
+          name: maskName,
+          uploadedBy: `${Collections.MICROSCOPISTS}/${labellerId}`,
+        }
+        const masks: Mask[] = [...(image.masks ?? []), mask]
+
+        const reachedSavedLimit: boolean = masks.length === SAVED_MASKS_TO_COMPLETE
+        const isCompleted: boolean = reachedSavedLimit
+        const status: ImageStatus = reachedSavedLimit ? ImageStatus.CONFIRMED_VALID : ImageStatus.IN_REVIEW
```

Comment:
Why is it needed this ImageStatus?

Reply:
It is intended to have the current status of the Image.
- It is described in the requirements, like [here](https://trello.com/c/CQvNhKe1/17-invalid-images-should-be-removed-from-data-set):

> ![image](https://user-images.githubusercontent.com/15014647/186372751-0cc4520e-3dad-4c9a-8fe0-7e3b5e888ca0.png)

- It makes sense to have a defined status because even though it is supposed to be possible to figure the "status" based on the other attributes (`isCompleted` + `markAsInvalid`/`masks`), these criteria might change.
For example, an Image with 4 masks is currently considered `CONFIRMED`, let's say that tomorrow the criteria changes: "now we need 5 masks to confirm an image", then all the already confirmed Images would be implicitly pending again.